File: src/contracts.ts
// Canonical API ↔ UI contracts.
// Imported by both the Hono server (server/src/) and the Vite UI (frontend/src/ via @contracts alias).
// Mirror: src/darwin/api/contracts.py — Pydantic counterpart; keep field names and types in sync.

// ─── Gene layers ──────────────────────────────────────────────────────────────

export type EmbeddingModel = 'gemini_3072' | 'gemini_1536' | 'gemini_768' | 'gemini_256';
export type QueryTransform = 'none' | 'hyde' | 'multi_query' | 'step_back';
export type RerankStrategy = 'none' | 'cross_encoder' | 'llm_rerank' | 'reciprocal_rank_fusion';

export interface RetrievalGenes {
  embedding_model: EmbeddingModel;
  chunk_size: number; // tokens: 128 | 256 | 512 | 1024
  chunk_overlap: number; // fraction: 0.0 | 0.1 | 0.25 | 0.5
  query_transform: QueryTransform;
  rerank: RerankStrategy;
  confidence_threshold: number; // [0.0, 1.0]
  source_routing: string[];
}

export type CoordinationProtocol = 'solo' | 'vote' | 'consult' | 'debate';
export type DisagreementResolver = 'majority_vote' | 'highest_confidence' | 'debate';

export interface CoordinationGenes {
  protocol: CoordinationProtocol;
  consultation_count: number;
  disagreement_resolver: DisagreementResolver;
  timeout_ms: number;
}

export type MemoryStrategy = 'none' | 'checkpoint' | 'full_history';
export type ContextCompression = 'none' | 'summarize' | 'truncate';

export interface DurabilityGenes {
  memory_strategy: MemoryStrategy;
  checkpoint_every_n_turns: number;
  context_compression: ContextCompression;
}

// ─── Fitness ──────────────────────────────────────────────────────────────────

export interface FitnessComponents {
  relevance: number; // [0.0, 1.0]
  accuracy: number; // [0.0, 1.0]
  latency_ms: number;
  cost_usd: number;
}

export interface FitnessSummary {
  composite: number; // [0.0, 1.0]
  last_updated: string; // ISO 8601
}

// ─── Genome ───────────────────────────────────────────────────────────────────

export type GenomeStatus = 'alive' | 'retired';

export interface GenomeSummary {
  id: string;
  generation: number;
  status: GenomeStatus;
  fitness: FitnessSummary;
  retrieval_genes: RetrievalGenes;
  coordination_genes: CoordinationGenes;
  durability_genes: DurabilityGenes;
}

export interface GenomeDetail extends GenomeSummary {
  parent_ids: string[];
  created_at: string; // ISO 8601
  mutation_log: string[];
}

// ─── Retrieval trace ──────────────────────────────────────────────────────────

export interface RetrievalTraceItem {
  chunk_id: string;
  score: number;
  position: number;
}

// ─── Generation summary ───────────────────────────────────────────────────────

export interface GenerationRecord {
  generation: number;
  population_size: number;
  best_fitness: number;
  mean_fitness: number;
  diversity_index: number;
  selection: string;
  crossover_rate: number;
  mutation_rate: number;
  created_at: string; // ISO 8601
}

export interface FitnessCurvePoint {
  generation: number;
  best: number;
  mean: number;
  diversity: number;
}

// ─── API response wrappers ────────────────────────────────────────────────────

export interface FitnessCurveResponse {
  series: FitnessCurvePoint[];
}

export interface GenerationsResponse {
  generations: GenerationRecord[];
}

// ─── Lineage (/lineage/{genome_id}) ──────────────────────────────────────────

export interface LineageNode {
  id: string;
  generation: number;
  parent_ids: string[];
  fitness: FitnessSummary;
  retrieval_genes: RetrievalGenes;
  is_champion: boolean;
}

export interface LineageResponse {
  genome_id: string;
  nodes: LineageNode[];
}

// ─── Champions (/champions) ───────────────────────────────────────────────────

export interface ChampionRecord {
  id: string;
  original_genome_id: string;
  peak_fitness: number;
  generations_alive: number;
  lineage: string[]; // parent_id chain
  retired_at: string | null; // ISO 8601
  genome: GenomeSummary;
}

// ─── SSE events (/events) ─────────────────────────────────────────────────────

export type EvolutionEventType =
  | 'generation.evolved'
  | 'genome.born'
  | 'genome.retired'
  | 'champion.promoted'
  | 'query.completed';

export interface EvolutionEvent {
  event_type: EvolutionEventType;
  generation: number;
  data: Record<string, unknown>; // payload varies by event_type
  timestamp: string; // ISO 8601
}

// ─── Request / Response shapes ────────────────────────────────────────────────

export interface QueryRequest {
  text: string;
}

export interface QueryResponse {
  run_id: string;
  answer: string;
  winning_genome: GenomeSummary;
  fitness: FitnessComponents;
  composite_fitness: number;
  retrieval_trace: RetrievalTraceItem[];
}

export interface PopulationResponse {
  genomes: GenomeSummary[];
  generation: number;
  population_size: number;
}

File: src/darwin/api/contracts.py
# contracts.py — canonical API data contract for Darwin backend↔frontend sync.
# Mirror: ui/lib/contracts.ts — TypeScript counterpart; keep field names and types in sync.
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Gene layers — three-layer genome (ARCHITECTURE.md §1, fleet frozen-decision)
# ---------------------------------------------------------------------------


class RetrievalGenes(BaseModel):
    embedding_model: Literal["gemini_3072", "gemini_1536", "gemini_768", "gemini_256"]
    chunk_size: int  # tokens: 128 | 256 | 512 | 1024
    chunk_overlap: float  # fraction: 0.0 | 0.1 | 0.25 | 0.5
    query_transform: Literal["none", "hyde", "multi_query", "step_back"]
    rerank: Literal["none", "cross_encoder", "llm_rerank", "reciprocal_rank_fusion"]
    confidence_threshold: float = Field(ge=0.0, le=1.0)
    source_routing: list[str]


class CoordinationGenes(BaseModel):
    protocol: Literal["solo", "vote", "consult", "debate"]
    consultation_count: int = Field(ge=0)
    disagreement_resolver: Literal["majority_vote", "highest_confidence", "debate"]
    timeout_ms: int = Field(ge=0)


class DurabilityGenes(BaseModel):
    memory_strategy: Literal["none", "checkpoint", "full_history"]
    checkpoint_every_n_turns: int = Field(ge=1)
    context_compression: Literal["none", "summarize", "truncate"]


# ---------------------------------------------------------------------------
# Fitness
# ---------------------------------------------------------------------------


class FitnessComponents(BaseModel):
    relevance: float = Field(ge=0.0, le=1.0)
    accuracy: float = Field(ge=0.0, le=1.0)
    latency_ms: float = Field(ge=0.0)
    cost_usd: float = Field(ge=0.0)


class FitnessSummary(BaseModel):
    composite: float = Field(ge=0.0, le=1.0)
    last_updated: datetime


# ---------------------------------------------------------------------------
# Genome
# ---------------------------------------------------------------------------


class GenomeSummary(BaseModel):
    id: str
    generation: int
    status: Literal["alive", "retired"]
    fitness: FitnessSummary
    retrieval_genes: RetrievalGenes
    coordination_genes: CoordinationGenes
    durability_genes: DurabilityGenes


class GenomeDetail(GenomeSummary):
    parent_ids: list[str]
    created_at: datetime
    mutation_log: list[str]


# ---------------------------------------------------------------------------
# Retrieval trace
# ---------------------------------------------------------------------------


class RetrievalTraceItem(BaseModel):
    chunk_id: str
    score: float
    position: int


# ---------------------------------------------------------------------------
# Generation summary (fitness-curve / generations endpoints)
# ---------------------------------------------------------------------------


class GenerationRecord(BaseModel):
    generation: int
    population_size: int
    best_fitness: float
    mean_fitness: float
    diversity_index: float
    selection: str
    crossover_rate: float
    mutation_rate: float
    created_at: datetime


class FitnessCurvePoint(BaseModel):
    generation: int
    best: float
    mean: float
    diversity: float


# ---------------------------------------------------------------------------
# Lineage (/lineage/{genome_id})
# ---------------------------------------------------------------------------


class LineageNode(BaseModel):
    id: str
    generation: int
    parent_ids: list[str]
    fitness: FitnessSummary
    retrieval_genes: RetrievalGenes
    is_champion: bool


class LineageResponse(BaseModel):
    genome_id: str
    nodes: list[LineageNode]


# ---------------------------------------------------------------------------
# Champions (/champions)
# ---------------------------------------------------------------------------


class ChampionRecord(BaseModel):
    id: str
    original_genome_id: str
    peak_fitness: float
    generations_alive: int
    lineage: list[str]  # parent_id chain
    retired_at: datetime | None
    genome: GenomeSummary


# ---------------------------------------------------------------------------
# SSE events (/events)
# ---------------------------------------------------------------------------


class EvolutionEvent(BaseModel):
    event_type: Literal[
        "generation.evolved",
        "genome.born",
        "genome.retired",
        "champion.promoted",
        "query.completed",
    ]
    generation: int
    data: dict  # payload varies by event_type
    timestamp: datetime


# ---------------------------------------------------------------------------
# Request / Response shapes for API routes
# ---------------------------------------------------------------------------


class QueryRequest(BaseModel):
    text: str


class QueryResponse(BaseModel):
    run_id: str
    answer: str
    winning_genome: GenomeSummary
    fitness: FitnessComponents
    composite_fitness: float
    retrieval_trace: list[RetrievalTraceItem]


class PopulationResponse(BaseModel):
    genomes: list[GenomeSummary]
    generation: int
    population_size: int
